MetaMask/metamask-extension
# Disable eth_sign by default, allow users to toggle it back on

## Explanation

After looking at various phishing kits ([and de-obfuscating](https://twitter.com/sniko_/status/1587882460976713728)), there are enormous trends between different kits that they all abuse `eth_sign` to ask users to inadvertently sign a transaction (as MetaMask UI is opaque to an on-chain transaction being signed). You can read more about the attack [here](https://officercia.medium.com/how-hackers-may-steal-your-ethers-and-why-does-eth-sign-function-matter-69497ee11593).

I'd estimate that the abuse of `eth_sign` is near equal to the abuse of Wyvern and Seaport contracts in a manner to steal NFTs through creating orders for 0 eth to a specific beneficiary.

In short: `eth_sign` has been consistently abused by phishers/scammers to steal funds and the case may be that only power-users who use very legacy dapps may need to still use `eth_sign`.

* Disable `eth_sign` by default - throw an `ethErrors.rpc.methodNotFound` exception
* Allow the user to enable `eth_sign` through "Advanced Settings"

RPC Method preferences are stored within `rpcMethodPreferences` in the store state, allow for us to develop this further if someone wants to lock down rpc calls into MetaMask

## Screenshots/Screencaps
![Screenshot 2023-01-20 at 17 12 30](https://user-images.githubusercontent.com/2313704/213761789-4656cd56-e2e6-4274-ab31-0b0fc03f3d11.png)

## Manual Testing Steps

1. Go to https://metamask.github.io/test-dapp/
2. Open Developer Tools
3. Click "Sign" button under the heading "Eth Sign"
4. See the exception thrown and no prompt to sign within MetaMask
5. Go into MetaMask > Settings > Advanced Settings
6. Toggle eth_sign to be enabled
7. `goto: 1`

## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
